# Exploratory Analysis of eBay Car Sales

## Introduction

The objective of this project is to explore the dataset of used cars from  _eBay Kleinanzeigen_, a classified section of the German eBay website.

The dataset contains a sample of 50,000 data points with details of the car advertised and some basic information advertisment and the listing. The dataset can be extracted [here](https://data.world/data-society/used-cars-data).

In [1]:
import numpy as np
import pandas as pd

In [2]:
autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [5]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

The dataset contains 50,000 rows and 20 columns.
There are 5 columns that contain a considerable amount of null values.
Information of dates are stored as string data type.

It contains detailed manufactured information of the car sold, information of advertisement and the offerings.

## Cleaning the dataset

### Cleaning the column names for ease of use for analysis

In [6]:
# checking the column names of the dataset
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [7]:
mapping_dict = {'yearOfRegistration':'registration_year',
               'monthOfRegistration':'registration_month',
               'notRepairedDamage':'unrepaired_damage',
               'dateCreated':'ad_created',
                'dateCrawled':'date_crawled',
               'offerType':'offer_type',
               'abtest':'ab_test',
                'vehicleType':'vehicle_type',
                'powerPS':'power_ps',
               'fuelType':'fuel_type',
               'nrOfPictures':'no_of_pictures',
               'postalCode':'postal_code',
               'lastSeen':'last_seen',}
autos.rename(columns=mapping_dict,inplace=True)
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'no_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

The step above is used to rename the column labels of the dataset for the ease of use for the data exploratory process later.

We have renamed the columns that contain camelcase (e.g. offerType) to snakecase (e.g. offer_type) and some other columns to specify the meaning of the columns (e.g. dateCreated to ad_created).

## Removing unuseful columns

In [8]:
# exploring the values in each column
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-16 21:50:53  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  no_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

In [9]:
# exploring the seller and offer_type columns
autos['seller'].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [10]:
autos['offer_type'].value_counts()

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

Both `seller` and `offer_type` columns contain only one single value that is different from the rest of the values, they do not seem useful in this exploratory study. Both columns are to be removed.

In [11]:
# exploring ab_test column
autos['ab_test'].value_counts()

test       25756
control    24244
Name: ab_test, dtype: int64

In [12]:
# exploring no_of_pictures column
autos['no_of_pictures'].value_counts()

0    50000
Name: no_of_pictures, dtype: int64

The column of `no_of_pictures` does not contain any useful information either because all the ads contained 0 pictures. It is to be removed.

In [13]:
# removing unuseful columns
autos.drop(['seller','offer_type','no_of_pictures'],axis=1,inplace=True)

In [14]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

    price  ab_test vehicle_type  registration_year    gearbox  power_ps  \
0  $5,000  control          bus               2004    manuell       158   
1  $8,500  control    limousine               1997  automatik       286   
2  $8,990     test    limousine               2009    manuell       102   
3  $4,350  control   kleinwagen               2007  automatik        71   
4  $1,350     test        kombi               2003    manuell         0   

    model   odometer  registration_month fuel_type       brand  \
0  andere  150,000km                   3       lpg     peugeot   
1     7er  150,000km                   6    benzin         bmw   
2    golf   70,000km                   7    benzin  volkswagen   
3  fortwo   70,000km                   6    benzin       smart   
4   focus  150,000km                   7    benzin        ford   

  unrepaired_damage           ad_created  postal_code            last_seen  
0              nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

Columns `seller`,`offer_type` and `no_of_pictures` have been removed from the dataset.

## Converting strings to numeric data types

### Converting `price` column

In [15]:
# exploring price column
autos['price'].value_counts()

$0         1421
$500        781
$1,500      734
$2,500      643
$1,200      639
           ... 
$898          1
$25,698       1
$5,298        1
$3,349        1
$18,890       1
Name: price, Length: 2357, dtype: int64

In [16]:
# removing special characters in price value
autos['price'] = autos['price'].str.replace('$','').str.replace(',','')

In [17]:
# converting data type to float
autos['price'] = autos['price'].astype(float)

In [18]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

### Converting `odometer` column

In [19]:
# exploraing odometer column
autos['odometer'].value_counts()

150,000km    32424
125,000km     5170
100,000km     2169
90,000km      1757
80,000km      1436
70,000km      1230
60,000km      1164
50,000km      1027
5,000km        967
40,000km       819
30,000km       789
20,000km       784
10,000km       264
Name: odometer, dtype: int64

In [20]:
# removing special characters in odometer column
autos['odometer'] = autos['odometer'].str.replace(',','').str.replace('km','')

In [21]:
# converting odometer to float data type
autos['odometer'] = autos['odometer'].astype(float)

In [22]:
autos['odometer'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer, dtype: float64

In [23]:
# renaming the column to indicate odometer_km
autos.rename(columns={'odometer':'odometer_km'},inplace=True)

In [24]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

    price  ab_test vehicle_type  registration_year    gearbox  power_ps  \
0  5000.0  control          bus               2004    manuell       158   
1  8500.0  control    limousine               1997  automatik       286   
2  8990.0     test    limousine               2009    manuell       102   
3  4350.0  control   kleinwagen               2007  automatik        71   
4  1350.0     test        kombi               2003    manuell         0   

    model  odometer_km  registration_month fuel_type       brand  \
0  andere     150000.0                   3       lpg     peugeot   
1     7er     150000.0                   6    benzin         bmw   
2    golf      70000.0                   7    benzin  volkswagen   
3  fortwo      70000.0                   6    benzin       smart   
4   focus     150000.0                   7    benzin        ford   

  unrepaired_damage           ad_created  postal_code            last_seen  
0              nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

In this section, we have converted two columns `price` and `odometer` from strings to float for use. In addition, `odometer` column is renamed to be `odometer_km` to indicate the unit of the values.

## Exploring the data

### Exploring odometer column

For `odometer_km` column:

In [25]:
autos['odometer_km'].unique().shape

(13,)

In [26]:
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [27]:
autos['odometer_km'].value_counts()

150000.0    32424
125000.0     5170
100000.0     2169
90000.0      1757
80000.0      1436
70000.0      1230
60000.0      1164
50000.0      1027
5000.0        967
40000.0       819
30000.0       789
20000.0       784
10000.0       264
Name: odometer_km, dtype: int64

In [28]:
# explore the odometer with value 150000
autos[autos['odometer_km'] == 150000.0]

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
...                    ...                                                ...   
49991  2016-03-06 15:25:19                                         Kleinwagen   
49993  2016-03-15 18:47:35           Audi_A3__1_8l__Silber;_schoenes_Fahrzeug   
49994  2016-03-22 17:36:42  Audi_A6__S6__Avant_4.2_quattro_eventuell_Tausc...   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

        price  ab_test vehicle_type  registration_year    gearbox  power_ps  \
0      5000.0  control          bus               2004    manuell       158   
1      8500.0  control    limousine               1997  automatik       286   
4      1350.0     test        kombi               2003    manuell         0   
5      7900.0     test          bus               2006  automatik       150   
6       300.0     test    limousine               1995    manuell        90   
...       ...      ...          ...                ...        ...       ...   
49991   500.0  control          NaN               2016    manuell         0   
49993  1650.0  control   kleinwagen               1997    manuell         0   
49994  5000.0  control        kombi               2001  automatik       299   
49996  1980.0  control       cabrio               1996    manuell        75   
49999  1250.0  control    limousine               1996    manuell       101   

         model  odometer_km  registration_month fuel_type       brand  \
0       andere     150000.0                   3       lpg     peugeot   
1          7er     150000.0                   6    benzin         bmw   
4        focus     150000.0                   7    benzin        ford   
5      voyager     150000.0                   4    diesel    chrysler   
6         golf     150000.0                   8    benzin  volkswagen   
...        ...          ...                 ...       ...         ...   
49991   twingo     150000.0                   0    benzin     renault   
49993      NaN     150000.0                   7    benzin        audi   
49994       a6     150000.0                   1    benzin        audi   
49996    astra     150000.0                   5    benzin        opel   
49999   vectra     150000.0                   1    benzin        opel   

      unrepaired_damage           ad_created  postal_code            last_seen  
0                  nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1                  nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
4                  nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50  
5                   NaN  2016-03-21 00:00:00        22962  2016-04-06 09:45:21  
6                   NaN  2016-03-20 00:00:00        31535  2016-03-23 02:48:59  
...                 ...                  ...          ...                  ...  
49991               NaN  2016-03-06 00:00:00        61350  2016-03-06 18:24:19  
49993               NaN  2016-03-15 00:00:00        65203  2016-04-06 19:46:53  
49994              nein  2016-03-22 00:00:00        46537  2016-04-06 08:16:39  
49996              nein  2016-03-28 00:00:00        44807  2016-04-02 14:18:02  
49999              nein  2016-03-13 00:00:00        45897  2016-04-06 21:18:48  

[32424 rows x 17 columns]

From preliminary observation, over 50% of the vehicle in the data set has the value of odometer of 150,000.

There seems to be no common characters of these vehicles.

### Exploring price column

For `price` column:

In [29]:
autos['price'].unique().shape

(2357,)

In [30]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [31]:
autos['price'].value_counts().sort_index()

0.0           1421
1.0            156
2.0              3
3.0              1
5.0              2
              ... 
10000000.0       1
11111111.0       2
12345678.0       3
27322222.0       1
99999999.0       1
Name: price, Length: 2357, dtype: int64

In [32]:
autos["price"].value_counts().sort_index(ascending=True).head(15)

0.0     1421
1.0      156
2.0        3
3.0        1
5.0        2
8.0        1
9.0        1
10.0       7
11.0       2
12.0       3
13.0       2
14.0       1
15.0       2
17.0       3
18.0       1
Name: price, dtype: int64

In [33]:
autos["price"].value_counts().sort_index(ascending=False).head(15)

99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
350000.0      1
345000.0      1
299000.0      1
295000.0      1
265000.0      1
Name: price, dtype: int64

From this exploration, we have observed the following:
- There are 2357 unique values in the column. 50% of the vehicle (1421 vehicles) are registered with the price of 0.
- There are also some outliners in the dataset. We can see that from price 350000 onwwards, the prices esculate to 3 - 300 times than this value. There are only in total 12 listings with these high values.

Since these outliners will affect our analysis later, we are going to remove them from our data set.

We are going to keep vehicles of prices between 1 - 35000 in this case.

In [34]:
autos = autos[autos['price'].between(1, 35000)]

autos['price'].describe()

count    48056.000000
mean      5331.427168
std       6047.229364
min          1.000000
25%       1200.000000
50%       2999.000000
75%       7200.000000
max      35000.000000
Name: price, dtype: float64

We have removed these price outliners (price equals 0 and price > $7200) in the data set and there are 34479 rows remaining.

### Exploring date information

In this section, we will explore the date information in the data set.

There are 5 columns that represent date values in the data set:
- `date_crawled`
- `last_seen`
- `ad_created`
- `registration_month`
- `registration_year`

In [35]:
autos.dtypes

date_crawled           object
name                   object
price                 float64
ab_test                object
vehicle_type           object
registration_year       int64
gearbox                object
power_ps                int64
model                  object
odometer_km           float64
registration_month      int64
fuel_type              object
brand                  object
unrepaired_damage      object
ad_created             object
postal_code             int64
last_seen              object
dtype: object

`date_crawled`, `ad_created` and `last_seen` columns are object type in the table.

In [36]:
autos[['date_crawled','ad_created','last_seen']][:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [37]:
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025366
2016-03-06    0.014067
2016-03-07    0.036083
2016-03-08    0.033378
2016-03-09    0.033149
2016-03-10    0.032233
2016-03-11    0.032483
2016-03-12    0.037082
2016-03-13    0.015690
2016-03-14    0.036603
2016-03-15    0.034293
2016-03-16    0.029653
2016-03-17    0.031630
2016-03-18    0.012860
2016-03-19    0.034689
2016-03-20    0.037872
2016-03-21    0.037144
2016-03-22    0.032920
2016-03-23    0.032233
2016-03-24    0.029278
2016-03-25    0.031526
2016-03-26    0.032379
2016-03-27    0.031047
2016-03-28    0.034793
2016-03-29    0.034189
2016-03-30    0.033752
2016-03-31    0.031775
2016-04-01    0.033607
2016-04-02    0.035438
2016-04-03    0.038684
2016-04-04    0.036582
2016-04-05    0.012922
2016-04-06    0.003184
2016-04-07    0.001415
Name: date_crawled, dtype: float64

It seems that the data is crawled almost everyday during a one-month period from 2016-03-05 to 2016-04-07.

In [38]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
                ...   
2016-04-03    0.038913
2016-04-04    0.036978
2016-04-05    0.011632
2016-04-06    0.003267
2016-04-07    0.001269
Name: ad_created, Length: 75, dtype: float64

The advertisements seem to be created over a period of 10 months since June 2015.

In [39]:
autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001082
2016-03-06    0.004370
2016-03-07    0.005390
2016-03-08    0.007491
2016-03-09    0.009655
2016-03-10    0.010758
2016-03-11    0.012465
2016-03-12    0.023930
2016-03-13    0.008948
2016-03-14    0.012631
2016-03-15    0.015981
2016-03-16    0.016460
2016-03-17    0.028196
2016-03-18    0.007346
2016-03-19    0.015919
2016-03-20    0.020643
2016-03-21    0.020663
2016-03-22    0.021475
2016-03-23    0.018603
2016-03-24    0.019810
2016-03-25    0.019269
2016-03-26    0.016814
2016-03-27    0.015690
2016-03-28    0.020996
2016-03-29    0.022474
2016-03-30    0.024908
2016-03-31    0.023826
2016-04-01    0.022911
2016-04-02    0.025012
2016-04-03    0.025283
2016-04-04    0.024659
2016-04-05    0.124001
2016-04-06    0.221346
2016-04-07    0.130993
Name: last_seen, dtype: float64

It seems that the dates in `last_seen` column, which represents the date that the ad is removed, is very similar to the dates in `date_crawled` column.

The last three dates in the `last_seen` column have also considerably higher percentage compared to the rest of the dates. However, this is no direct indication in the car sales. It could be possible that the ads were removed before the vehicles were sold.

In [40]:
autos['registration_year'].describe()

count    48056.000000
mean      2004.635862
std         87.023293
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

In [41]:
autos['registration_year'].value_counts().sort_index()

1000    1
1001    1
1111    1
1800    2
1910    5
       ..
5000    4
5911    1
8888    1
9000    1
9999    3
Name: registration_year, Length: 92, dtype: int64

From the observation, there are some values of the `registration_year` that seem to be incorrect, e.g. years before 1900 and after 2016 (which is the year of the data being crawled).

## Handling incorrect registration year data

In this section, we are going to handle the incorrect values in the `registration_year` column.

Since the data being crawled is 2016, any registration year value after 2016 should be invalid. In addition, any registration year value before 1900 should also be invalid because of the beginning of manufacture years of vehicles.

Therefore, the valid values for the registration year should be within the interval of: 1900-2016

In [42]:
autos["registration_year"].between(1900,2016).value_counts() / len(autos["registration_year"])

True     0.960941
False    0.039059
Name: registration_year, dtype: float64

Data of vehicles which is out of the interval of registration year is ~4% of the overall data set.

Since it is not a significant amount of data, it is safe to remove such data in this case.

In [43]:
autos = autos[autos["registration_year"].between(1900,2016)]

In [44]:
autos["registration_year"].between(1900,2016).value_counts()

True    46179
Name: registration_year, dtype: int64

In this section, we have removed the data which does not have the registration year between the defined interval of 1900-2016 from the data set. There are 48028 vehicles remaining after the removal.

## Exploring vehicle brands

In this section, we are going to explore the brands of vehicles in the data set.

In [45]:
autos['brand'].value_counts(normalize=True)

volkswagen        0.212867
bmw               0.109769
opel              0.108729
mercedes_benz     0.095498
audi              0.085299
ford              0.070443
renault           0.047641
peugeot           0.030165
fiat              0.025921
seat              0.018472
skoda             0.016588
nissan            0.015440
mazda             0.015332
smart             0.014314
citroen           0.014162
toyota            0.012820
hyundai           0.010134
sonstige_autos    0.009247
volvo             0.009247
mini              0.008857
mitsubishi        0.008294
honda             0.007904
kia               0.007146
alfa_romeo        0.006691
suzuki            0.005998
chevrolet         0.005717
chrysler          0.003551
porsche           0.003378
dacia             0.002664
daihatsu          0.002534
subaru            0.002165
jeep              0.002144
land_rover        0.001754
saab              0.001667
jaguar            0.001516
daewoo            0.001516
trabant           0.001408
r

We can look at brands that represent more than 5% of the listings.

In [46]:
brand_perc = autos["brand"].value_counts(normalize=True)
top_brands = brand_perc[brand_perc > 0.05].index
print(top_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [48]:
brands_mean_price = {}

for brand in top_brands:
    brand_row = autos[autos['brand'] == brand]
    mean_price = brand_row['price'].mean()
    brands_mean_price[brand] = int(mean_price)
    
brands_mean

{'volkswagen': 5280,
 'bmw': 7742,
 'opel': 2968,
 'mercedes_benz': 7727,
 'audi': 8377,
 'ford': 3577}

From the observation, there are 6 vehicle brands at least 5% of listings in the data set. In addition, we can see that there are a variety of vehicle class in the listings:

- Luxury brands: Audi, Mercedes Benz, BMW
- Middle-class brands: Volkswagen
- Budget brands: Ford, Opel

## Exploring mileage of top 6 brands

From the top 6 car brands, we have observed that there is a distinct price gap. In this section, we will explore the average mileage for these cars and see if there's any visible link with mean price.

In [52]:
#calculate the mean mileage for each top brand
brand_mean_mileage = {}

for brand in top_brands:
    brand_row = autos[autos['brand'] == brand]
    mean_mileage = brand_row['odometer_km'].mean()
    brand_mean_mileage[brand] = int(mean_mileage)
    
brand_mean_mileage

{'volkswagen': 128961,
 'bmw': 133689,
 'opel': 129310,
 'mercedes_benz': 132235,
 'audi': 131315,
 'ford': 124495}

In [53]:
#converting both dictionaries to series objects
bmp_series = pd.Series(brands_mean_price)
bmm_series = pd.Series(brand_mean_mileage)

In [55]:
#creating a dataframe from the series
df_brand = pd.DataFrame(bmp_series, columns=['mean_price'])
df_brand['mean_mileage'] = bmm_series

df_brand

mean_price  mean_mileage
volkswagen           5280        128961
bmw                  7742        133689
opel                 2968        129310
mercedes_benz        7727        132235
audi                 8377        131315
ford                 3577        124495

From the analysis, we have observed that:

1. Prices of the cars seem to be positively correlated to the average mileages of the cars, yet there is one exception: Opel

2. Although Opel is among the lowest priced compared to the rest of the brands, the average mileage of Opel cars are higher than that of Volkswage and Ford, which are priced higher